cli: introduce exclude id list for extension-packs

vince-fugnitto opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

<!--
Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See SECURITY.md at the root of this repository, to learn how to report
vulnerabilities.
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->

The pull-request introduces a mechanism to exclude extensions by `id` when resolving builtin **extension-packs**. The purpose is to allow application developers the flexibility to use extension-packs but still have the control to exclude problematic or unwanted extensions that may be referenced by the pack. Previously, post-processing would be necessary to remove such extensions, for example by deleting the plugin in the plugins directory before a build.

The changes introduce `theiaPluginsExcludeIds` which lists extensions by `id` and is used to exclude extensions when resolving extension-packs. 

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

- update the [`package.json`](https://github.com/eclipse-theia/theia/blob/471f10ae5beb24f7b072e80acf0e8b22c31e58ba/package.json#L87) to the following:

```json
"theiaPlugins": {
  "pack": "https://github.com/vince-fugnitto/vscode-builtin-extensionPack/releases/download/0.0.1/builtin-extension-pack-1.45.1.vsix"
},
"theiaPluginsExcludeIds": [
  "vscode.cpp"
]
```

- execute the cli command `yarn download:plugins`
- confirm that the pack is downloaded, and individual extensions are resolved
- confirm that `vscode.cpp` is logged as skipped
- confirm that `vscode.cpp` is not present under the `plugins` folder

_example:_

![image](https://user-images.githubusercontent.com/40359487/130655941-af0a0413-3556-4247-84fd-8c4cf7ca2ce1.png)


#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: vince-fugnitto <vincent.fugnitto@ericsson.com>

marcdumais-work commented:
Looks nice, thanks @vince-fugnitto .

Looking at how this works, I am reminded of another use-case, beyond extension packs. Let's say that there is an extension you like but for some reasons it's not working well ATM, or an extension already in `package.json` stops working correctly, the only way forward ATM is to not have it mentioned in `package.json`. 

If this PR was made slightly more generic, so that the exclusion works when downloading extensions, not just extension packs, one could instead add it or keep it, and also exclude it, until it works as wanted. Then one needs only remove the exclusion instead of thinking adding the extension. 

WDYT?

alvsan09 reviewed (commented):
I have tested the `CLI` and validated that the plugins are successfully excluded.
![exclude_extensions_from_pack](https://user-images.githubusercontent.com/76971376/130693796-20b7e05f-eabd-4091-9bba-20463c21b475.png)

However, if I then try to start the IDE, I can see in the log entries that those plugins are downloaded to the `tmp/..` folder and loaded from there, see: 
![deploying_plugins_](https://user-images.githubusercontent.com/76971376/130693992-ad6b26fe-286a-4dc4-9bce-94e733dc16de.png)

marcdumais-work commented on dev-packages/cli/README.md:
````diff
@@ -318,6 +318,15 @@ The property `theiaPlugins` describes the list of plugins to download, for examp
+The `ids` referenced by the property will not be downloaded when resolving extension-packs, and can be used to omit extensions which are problematic or unwanted. The format of the property is as follows:
+
+```json
+"theiaPluginsExcludeIds": [
````
Maybe we can simplify and use "theiaPluginsExcluded"? It seems that excluding by ID is the logical, intuitive way to go and you documented it too.

marcdumais-work replied:
Maybe we can simplify and use "theiaPluginsExcluded"? It seems that excluding by ID is the logical, intuitive way to go and you documented it too.

vince-fugnitto commented:
> Looking at how this works, I am reminded of another use-case, beyond extension packs. Let's say that there is an extension you like but for some reasons it's not working well ATM, or an extension already in `package.json` stops working correctly, the only way forward ATM is to not have it mentioned in `package.json`.
> 
> If this PR was made slightly more generic, so that the exclusion works when downloading extensions, not just extension packs, one could instead add it or keep it, and also exclude it, until it works as wanted. Then one needs only remove the exclusion instead of thinking adding the extension.
> 
> WDYT?

@marcdumais-work I was thinking of that use-case but I did not see a reason as to why they would list an extension under `theiaPlugins` but want to exclude it as part of `theiaPluginsExcludeIds`. It would make more sense to me to simply not reference it in the first place.

Also, we use a `arbitrary name: url` for `theiaPlugins` mapping so it means that we would first need to download the plugins under then delete them off the filesystem if we scan the extensions and see an `id` in the exclude lists.

marcdumais-work commented:
> However, if I then try to start the IDE, I can see in the log entries that those plugins are downloaded to the `tmp/..` folder and loaded from there, see:
> ![deploying_plugins_](https://user-images.githubusercontent.com/76971376/130693992-ad6b26fe-286a-4dc4-9bce-94e733dc16de.png)

Good catch! I guess at runtime we resolve the pack again, and helpfully download anything that might have been missed at build-time :)

marcdumais-work commented:
> Also, we use a `arbitrary name: url` for `theiaPlugins` mapping so it means that we would first need to download the plugins under then delete them off the filesystem if we scan the extensions and see an `id` in the exclude lists.

I have been thinking that the way we specify plugins by URL is a bit dated - we should maybe support by id, but that's for another PR.

paul-marechal commented on dev-packages/cli/README.md:
```diff
@@ -318,6 +318,15 @@ The property `theiaPlugins` describes the list of plugins to download, for examp
 }
```
Would a good candidate be the builtins extension pack if/once available? Otherwise I don't see much value in it.

vince-fugnitto commented:
> > Also, we use a `arbitrary name: url` for `theiaPlugins` mapping so it means that we would first need to download the plugins under then delete them off the filesystem if we scan the extensions and see an `id` in the exclude lists.
> 
> I have been thinking that the way we specify plugins by URL is a bit dated - we should maybe support by id, but that's for another PR.

@marcdumais-work we could, but it means that we can no longe reference a specific version of a builtin anymore. An application may choose to support a specific builtin version, and not have the framework resolve it to the latest API.

vince-fugnitto commented:
@marcdumais-work @alvsan09 the pull-request is ready for another review.
I updated the functionality based on feedback (and refactored the methods for handling extension-packs) to hopefully make it clearer.

I also fixed the issue were extension-packs referenced as builtins are resolved at runtime by removing them from the plugins folder after we complete the resolvement and download of the `ids` they reference at buildtime.
